solidity/ecdsa/contracts/WalletFactory.sol

```diff
@@ -0,0 +1,315 @@
+    ///         details.
+    /// @dev The message to be signed by each member is keccak256 hash of the
+    ///      calculated group public key, misbehaved members indices and DKG
+    ///      start block. The calculated hash should be prefixed with prefixed with
+    ///      `\x19Ethereum signed message:\n` before signing, so the message to
+    ///      sign is:
+    ///      `\x19Ethereum signed message:\n${keccak256(groupPubKey,misbehavedIndices,startBlock)}`
+    /// @param dkgResult DKG result.
+    function submitDkgResult(DKG.Result calldata dkgResult) external {
+        dkg.submitResult(dkgResult);
+
+        address clonedWalletAddress = createClone(address(masterWallet));
+        require(
+            clonedWalletAddress != address(0),
+            "Cloned wallet address is 0"
+        );
+
+        emit WalletCreated(
+            clonedWalletAddress,
+            keccak256(abi.encode(dkgResult))
+        );
+
+        // We expect `dkg.submitResult` function verifies the current state of
+        // the wallet creation process. It is expected that at this point the
+        // wallet reference is not set, hence we won't be overwriting any previous
+        // value.
+        wallet = Wallet(clonedWalletAddress);
+
+        wallet.init(
+            address(this),
+            hashGroupMembers(
+                dkgResult.members,
+                dkgResult.misbehavedMembersIndices
+            ),
+            keccak256(dkgResult.groupPubKey)
+        );
+    }
+
+    /// @notice Notifies about DKG timeout. Pays the sortition pool unlocking
+    ///         reward to the notifier.
+    function notifyDkgTimeout() external {
+        dkg.notifyTimeout();
+
+        // TODO: Implement transferDkgRewards
+        // transferDkgRewards(msg.sender, sortitionPoolUnlockingReward);
+
+        //slither-disable-next-line reentrancy-benign
+        dkg.complete();
+
+        delete wallet;
+    }
+
+    /// @notice Approves DKG result. Can be called when the challenge period for
+    ///         the submitted result is finished. Considers the submitted result
+    ///         as valid, pays reward to the approver, bans misbehaved group
+    ///         members from the sortition pool rewards, and completes the group
+    ///         creation by activating the candidate group. For the first
+    ///         `resultSubmissionEligibilityDelay` blocks after the end of the
+    ///         challenge period can be called only by the DKG result submitter.
+    ///         After that time, can be called by anyone.
+    /// @dev It transfers Wallet's ownership to the Wallet Factory owner.
+    /// @param dkgResult Result to approve. Must match the submitted result
+    ///        stored during `submitDkgResult`.
+    function approveDkgResult(DKG.Result calldata dkgResult) external {
+        uint32[] memory misbehavedMembers = dkg.approveResult(dkgResult);
+
+        emit WalletActivated(address(wallet));
```

Comment:
I think we can remove this event and save gas. Inside of dkg.approveResult we already have `emit DkgResultApproved(submittedResultHash, sender)`. And during the `submitDkgResult()` we `emit WalletCreated(walletAddress, submittedResultHash)` and based on hash we can fetch the wallet address. 
I agree that it is more readable to emit the address of an approved wallet, but because we pay extra attention on the gas usage, I think we can get rid of WalletActivated event.